backend/api/src/realms.rs

```diff
@@ -25,73 +38,126 @@ impl Realm {
         &self.name
     }
 
-    fn parent_id(&self) -> Id {
-        Id::realm(self.parent_key)
+    fn parent_id(&self) -> Option<Id> {
+        self.parent_key.map(Id::realm)
+    }
+
+    fn path(&self, context: &Context) -> String {
+        let mut segments = self.walk_up(context)
+            .map(|realm| &*realm.path_segment)
+            .collect::<Vec<_>>();
+        segments.reverse();
+        segments.join("/")
     }
 
-    fn parent(&self, context: &Context) -> &Realm {
-        &context.realm_tree.realms[&self.parent_key]
+    fn parent(&self, context: &Context) -> Option<&Realm> {
+        self.parent_key.map(|parent_key| &context.realm_tree.realms[&parent_key])
+    }
+
+    fn parents(&self, context: &Context) -> Vec<&Realm> {
+        let mut parents = self.walk_up(context).skip(1).collect::<Vec<_>>();
+        parents.reverse();
+        parents
     }
 
     fn children(&self, context: &Context) -> Vec<&Realm> {
-        context.realm_tree.children.get(&self.key)
-            .map(|children| {
-                children.iter()
-                    .map(|child_key| &context.realm_tree.realms[&child_key])
-                    .collect()
-            })
-            .unwrap_or_default()
+        self.child_keys.iter()
+            .map(|child| &context.realm_tree.realms[&child])
+            .collect()
     }
 }
 
 pub(crate) struct Tree {
     pub(crate) realms: HashMap<u64, Realm>,
-    children: HashMap<u64, Vec<u64>>,
+    from_path: HashMap<String, u64>,
 }
 
 impl Tree {
-    pub(crate) async fn raw_from_db(
-        db: &Pool,
-    ) -> anyhow::Result<impl TryStream<Ok = Realm, Error = impl std::error::Error>> {
-        let row_stream = db.get().await?
-            .query_raw(
-                "select id, name, parent from realms",
-                std::iter::empty(),
-            ).await?;
-
-        Ok(row_stream.map_ok(|row| Realm {
-            key: row.get_key(0),
-            name: row.get(1),
-            parent_key: row.get_key(2),
-        }))
-    }
-
     pub(crate) async fn from_db(db: &Pool) -> anyhow::Result<Self> {
         // We store the nodes of the realm tree in a hash map
         // accessible by the database ID
-        let realms = Self::raw_from_db(db).await?
+        let mut realms = db.get().await?
+            .query_raw(
+                "select id, name, parent, path_segment from realms",
+                std::iter::empty(),
+            ).await?
```

Comment:
```suggestion
            )
            .await?
```

Mhhh... I think?

Reply:
I'd prefer not to? :sweat_smile: 

We talked about this before, but I pretty much never want `.await?` on a line by itself, because that draws too much attention to incidental complexity.

Reply:
rustfmt says so. It even puts `.get()` and the first `.await?` in its own line. And I don't think it's that bad really. It's just one additional line. I think it makes sense to expect all `.something` "items" in one of these chains to go on its own line.

Reply:
We don't have it like this everywhere, anyway, though. x) Dunno, I don't like it. If you insist, I can begrudgingly change it, but if we defer to `rustfmt` like this, why don't we just use it? :thinking:

Reply:
> We don't have it like this everywhere

Examples? Like where there is a . chain that is not just in one line and where `.await` is not on its own?

And regarding rustfmt: because rustfmt sometimes formats in a way that we both disagree with. We said that we use the standard style if we can't decide, right? And formatting every . element in a chain in its own line is just the standard Rust formatting, yes.

Reply:
Fiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiiine.